### ISSUE TITLE
SIP <-> record revision

### ISSUE THREAD

**Issue by kaplun** [inveniosoftware/invenio-sipstore#3]
In INSPIRE, (see also the json-merger project) we are looking for a SIP store such as this one, in order to automatize merging and updating of records.

However let's say we have a record with this history:
1) record is harvested from arXiv (arXiv version stored in SIP store)
2) record is curated by a human
3) record arrives also from a publishers and its content is merged (publisher version stored in SIP store)
3) new revision from arXiv arrives and is merged with the record (while respecting information from cataloguer and golden truth provided by the publisher).

How can we capture the fact that the given record is the result of the merging from several sources? In our case we would have that the current record is the result of a sequence of merges from different sources whose information is stored in several SIP packages (one from arXiv, one from publisher).
Somehow we would need to record the fact that version x of the record is the result of merging the record HEAD with the SIP arXiv.1, while version y is the result of merging the record HEAD with the SIP publisher.1.

What do you suggest? Is there a space for storing the association of a given SIP with a given history element of a record? Should this be stored here or in an extension of invenio-records models?

**Comment by kaplun** [comment]
Moved from @2. @lnielsen replied:
[...]
@kaplun Right now, we're just creating a basic model so we can migrate existing deposits.

I'm not sure where I would store it, but if you were to point from SIP -> Record instead of instead of SIP -> PID, you could just add the record revision as a result of the SIP.

Alternatively, you could perhaps add this information in the archival package, which is created after the SIP has been merged.
[...]

**Comment by tiborsimko** [comment]
> How can we capture the fact that the given record is the result of the merging from several sources? 

The answer is "field provenance" in my eyes.  In your example, there is only one record that has seen many SIPs in its lifetime that all contributed to its current metadata.  One field comes from this SIP, another field comes from that SIP, etc.  Which is a matter of the field provenance information that can be expressed in the current `SIP -> PID -> UUID` model, without going for a different `SIP -> UUID` model.  WDYT?

**Comment by kaplun** [comment]
@tiborsimko not really because:
* where would you stop attaching provenance info? at subfields level? If you have a field of type boolean, would you attach also the provenance to it? Is this going to be overkill?)
* every field is going to be the result of field merging manipulation: at that point you can't necessarily know that the resulting field come from a given SIP, but simply that, by merging the then-current HEAD with the incoming SIP you obtained that field. 

Basically you can imagine the history of a record as been linear, where at any given point a SIP was merged into into it, to compose the new revision. 

With the algorithm we are implementing for automatic merging of SIPs we need to be able to answer to questions such as: *give me the last arXiv version that was merged into this record*.

P.s. see also:
https://github.com/inspirehep/json-merger/issues/5
https://github.com/inspirehep/inspire-next/issues/1091

**Comment by kaplun** [comment]
@lnielsen:
> Alternatively, you could perhaps add this information in the archival package, which is created after the SIP has been merged.

This seems indeed inline with the above, i.e. augmenting every history element of a record with metadata referring to the SIP where this particular entry was derived from.

**Comment by tiborsimko** [comment]
> where would you stop attaching provenance info? 

That would be upon each installation to decide which fields need to be under "provenance control".

>  every field is going to be the result of field merging manipulation

This could lead to quite complex usage patterns indeed.  I hope one could stop at "subfield value provenance" and not go into deciphering "substring provenance" that a subfield value may be composed of.  Anyhow, regarding records and SIPs:

> give me the last arXiv version that was merged into this record

Since the arXiv input is happening in a format that requires massaging before it is turned to record JSON, and since this massaging can go wrong due to a bug or something, I'd advocate considering the separation of SIP concerns and record tracking concerns by having each such important source to create independent (and invisible) records too. This is very similar to what ADS does with separate arXiv versions and paper versions.  

Disadvantage: the record store would be much larger, but this can be solved by namespacing and sharding. 

Advantage: the common record machinery could then be used to fork, merge and otherwise work with records.  Note that with this approach, SIPs would be used only for storing exact arXiv input blobs as they were received by the harvester, before they were turned into record JSON. This is the primary purpose of SIPs, so that records be reproduced in case the arXiv XML -> record JSON manipulations change, for example.  Otherwise said, the SIPs would play somewhat "secondary" role, restricted to capturing all the input, while all the intelligent field merging and whatnot would happen later on the record level, as one would do for dupes for example. 

```
         SIP
          ^
          |
          |
arXiv -> XML -> convert -> internal record JSON      \
                                                       --> merge -> public record JSON
Springer -> XML -> convert -> internal record JSON   /
             |
             |
             V
            SIP
```

 This would separate nicely all the concerns for me: versioning of incoming sources, SIPs used for OAIS/SIP purposes, and records used for intelligent record merging purposes.

**Comment by lnielsen** [comment]
>  create independent (and invisible) records

This sounds to me to the best solution in terms of book the keeping. Records keep track of each revision, hence it's easy to point to the exact revisions of records which was used for a given merge.

UI-wise I'm not sure how it should work though given that a cataloger most likely would want to edit the merged record, but perhaps it can be "faked" :-)
